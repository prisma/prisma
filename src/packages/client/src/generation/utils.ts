import indent from 'indent-string'
import path from 'path'
import { DMMFClass } from '../runtime/dmmf'
import { DMMF } from '../runtime/dmmf-types'

export enum Projection {
  select = 'select',
  include = 'include',
}

export function getScalarsName(modelName: string): string {
  return `${modelName}Scalars`
}

export function getPayloadName(modelName: string): string {
  return `${modelName}GetPayload`
}

// export function getExtractName(modelName: string, projection: Projection) {
//   return `Extract${modelName}${capitalize(projection)}`
// }

export function getSelectName(modelName: string): string {
  return `${modelName}Select`
}

export function getAggregateName(modelName: string): string {
  return `Aggregate${capitalize(modelName)}`
}

export function getAvgAggregateName(modelName: string): string {
  return `${capitalize(modelName)}AvgAggregateOutputType`
}

export function getSumAggregateName(modelName: string): string {
  return `${capitalize(modelName)}SumAggregateOutputType`
}

export function getMinAggregateName(modelName: string): string {
  return `${capitalize(modelName)}MinAggregateOutputType`
}

export function getMaxAggregateName(modelName: string): string {
  return `${capitalize(modelName)}MaxAggregateOutputType`
}

export function getAggregateInputType(aggregateOutputType: string): string {
  return aggregateOutputType.replace(/OutputType$/, 'InputType')
}

export function getAggregateArgsName(modelName: string): string {
  return `Aggregate${capitalize(modelName)}Args`
}

export function getAggregateGetName(modelName: string): string {
  return `Get${capitalize(modelName)}AggregateType`
}

export function getAggregateScalarGetName(modelName: string): string {
  return `Get${capitalize(modelName)}AggregateScalarType`
}

export function getIncludeName(modelName: string): string {
  return `${modelName}Include`
}

export function getDefaultName(modelName: string): string {
  return `${modelName}Default`
}

export function getFieldArgName(field: DMMF.SchemaField): string {
  return getArgName(
    (field.outputType.type as DMMF.OutputType).name,
    field.outputType.isList,
  )
}

export function getArgName(name: string, isList: boolean): string {
  if (!isList) {
    return `${name}Args`
  }

  return `FindMany${name}Args`
}

// we need names for all top level args,
// as GraphQL doesn't have the concept of unnamed args
export function getModelArgName(
  modelName: string,
  action?: DMMF.ModelAction,
): string {
  if (!action) {
    return `${modelName}Args`
  }
  switch (action) {
    case DMMF.ModelAction.findMany:
      return `FindMany${modelName}Args`
    case DMMF.ModelAction.findOne:
      return `FindOne${modelName}Args`
    case DMMF.ModelAction.findFirst:
      return `FindFirst${modelName}Args`
    case DMMF.ModelAction.upsert:
      return `${modelName}UpsertArgs`
    case DMMF.ModelAction.update:
      return `${modelName}UpdateArgs`
    case DMMF.ModelAction.updateMany:
      return `${modelName}UpdateManyArgs`
    case DMMF.ModelAction.delete:
      return `${modelName}DeleteArgs`
    case DMMF.ModelAction.create:
      return `${modelName}CreateArgs`
    case DMMF.ModelAction.deleteMany:
      return `${modelName}DeleteManyArgs`
  }
}

export function getDefaultArgName(
  dmmf: DMMFClass,
  modelName: string,
  action: DMMF.ModelAction,
): string {
  const mapping = dmmf.mappings.find((m) => m.model === modelName)!

  const fieldName = mapping[action]
  const operation = getOperation(action)
  const queryType = operation === 'query' ? dmmf.queryType : dmmf.mutationType
  const field = queryType.fields.find((f) => f.name === fieldName)!
  return (field.args[0].inputTypes[0].type as DMMF.InputType).name
}

export function getOperation(action: DMMF.ModelAction): 'query' | 'mutation' {
  if (
    action === DMMF.ModelAction.findMany ||
    action === DMMF.ModelAction.findOne
  ) {
    return 'query'
  }
  return 'mutation'
}

/**
 * Used to render the initial client args
 * @param modelName
 * @param fieldName
 * @param mapping
 */
export function renderInitialClientArgs(
  actionName: DMMF.ModelAction,
  fieldName: string,
  mapping: DMMF.Mapping,
): string {
  return `
  dmmf,
  fetcher,
  '${getOperation(actionName)}',
  '${fieldName}',
  '${mapping.plural}.${actionName}',
  args || {},
  [],
  errorFormat,
  measurePerformance\n`
}

export function getFieldTypeName(field: DMMF.SchemaField): string {
  if (typeof field.outputType.type === 'string') {
    return field.outputType.type
  }

  return field.outputType.type.name
}

export function getType(
  name: string,
  isList: boolean,
  isOptional?: boolean,
): string {
  return name + (isList ? '[]' : '') + (isOptional ? ' | null' : '')
}

export function getFieldType(field: DMMF.SchemaField): string {
  return getType(getFieldTypeName(field), field.outputType.isList)
}

interface SelectReturnTypeOptions {
  name: string
  actionName: DMMF.ModelAction
  renderPromise?: boolean
  hideCondition?: boolean
  isField?: boolean
  fieldName?: string
  projection: Projection
}

/**
 * Get the complicated extract output
 * @param name Model name
 * @param actionName action name
 */
export function getSelectReturnType({
  name,
  actionName,
  renderPromise = true,
  hideCondition = false,
  isField = false, // eslint-disable-line @typescript-eslint/no-unused-vars
}: SelectReturnTypeOptions): string {
  const isList = actionName === DMMF.ModelAction.findMany

  if (actionName === 'deleteMany' || actionName === 'updateMany') {
    return `Promise<BatchPayload>`
  }

  /**
   * Important: We handle findMany or isList special, as we don't want chaining from there
   */
  if (isList || hideCondition) {
    const listOpen = isList ? 'Array<' : ''
    const listClose = isList ? '>' : ''
    const promiseOpen = renderPromise ? 'Promise<' : ''
    const promiseClose = renderPromise ? '>' : ''

    return `CheckSelect<T, ${promiseOpen}${listOpen}${name}${listClose}${promiseClose}, ${promiseOpen}${listOpen}${getPayloadName(
      name,
    )}<T>${listClose}${promiseClose}>`
  }

  return `CheckSelect<T, Prisma__${name}Client<${getType(name, isList)}${actionName === 'findOne' ? ' | null' : ''
    }>, Prisma__${name}Client<${getType(getPayloadName(name) + '<T>', isList)}${actionName === 'findOne' ? ' | null' : ''
    }>>`
}

export function isQueryAction(
  action: DMMF.ModelAction,
  operation: 'query' | 'mutation',
): boolean {
  if (!(action in DMMF.ModelAction)) {
    return false
  }
  const result =
    action === DMMF.ModelAction.findOne || action === DMMF.ModelAction.findMany
  return operation === 'query' ? result : !result
}

export function capitalize(str: string): string {
  return str[0].toUpperCase() + str.slice(1)
}

export function indentAllButFirstLine(
  str: string,
  indentation: number,
): string {
  const lines = str.split('\n')

  return lines[0] + '\n' + indent(lines.slice(1).join('\n'), indentation)
}

export function getRelativePathResolveStatement(
  outputDir: string,
  cwd?: string,
): string {
  if (!cwd) {
    return 'undefined'
  }
  return `path.resolve(__dirname, ${JSON.stringify(
    path.relative(outputDir, cwd),
  )})`
}

function flatten(array): any[] {
  return Array.prototype.concat.apply([], array)
}

export function flatMap<T, U>(
  array: T[],
  callbackFn: (value: T, index: number, array: T[]) => U[],
  thisArg?: any,
): U[] {
  return flatten(array.map(callbackFn, thisArg))
}
