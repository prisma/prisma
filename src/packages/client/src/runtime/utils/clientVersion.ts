export const clientVersion = require('../../../package.json').version