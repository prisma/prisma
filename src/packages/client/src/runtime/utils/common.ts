import chalk from 'chalk'
import indent from 'indent-string'
import leven from 'js-levenshtein'
import { DMMF } from '../dmmf-types'
import Decimal from 'decimal.js'

export interface Dictionary<T> {
  [key: string]: T
}

export const keyBy: <T>(collection: T[], prop: string) => Dictionary<T> = (
  collection,
  prop,
) => {
  const acc = {}

  for (const obj of collection) {
    const key = obj[prop]
    acc[key] = obj
  }
  return acc
}

export const keyBy2: <T>(
  collection1: T[],
  collection2: T[],
  prop: string,
) => Dictionary<T> = (collection1, collection2, prop) => {
  const acc = {}

  for (const obj of collection1) {
    const key = obj[prop]
    acc[key] = obj
  }

  for (const obj of collection2) {
    const key = obj[prop]
    acc[key] = obj
  }
  return acc
}

export const ScalarTypeTable = {
  String: true,
  Int: true,
  Float: true,
  Boolean: true,
  Long: true,
  DateTime: true,
  ID: true,
  UUID: true,
  Json: true,
  Bytes: true,
  Xml: true,
  Decimal: true
}

export function isScalar(str: string): boolean {
  if (typeof str !== 'string') {
    return false
  }
  return ScalarTypeTable[str] || false
}

export const GraphQLScalarToJSTypeTable = {
  String: 'string',
  Int: 'number',
  Float: 'number',
  Boolean: 'boolean',
  Long: 'number',
  DateTime: ['Date', 'string'],
  ID: 'string',
  UUID: 'string',
  Json: 'JsonValue',
  Bytes: 'Buffer',
  Xml: 'string',
  Decimal: 'Decimal'
}

export const JSOutputTypeToInputType = {
  JsonValue: 'InputJsonValue',
}

export const JSTypeToGraphQLType = {
  string: 'String',
  boolean: 'Boolean',
  object: 'Json',
}

export function stringifyGraphQLType(
  type: string | DMMF.InputType | DMMF.SchemaEnum,
) {
  if (typeof type === 'string') {
    return type
  }
  return type.name
}

export function wrapWithList(str: string, isList: boolean) {
  if (isList) {
    return `List<${str}>`
  }

  return str
}

export function getGraphQLType(
  value: any,
  potentialType?: string | DMMF.SchemaEnum | DMMF.InputType,
): string {
  if (value === null) {
    return 'null'
  }

  // https://github.com/MikeMcl/decimal.js/blob/master/decimal.js#L4499
  if (Decimal.isDecimal(value)) {
    return 'Decimal'
  }

  if (Buffer.isBuffer(value)) {
    return 'Bytes'
  }

  if (potentialType && potentialType.toString() === 'Xml' && typeof value === 'string') {
    return 'Xml'
  }

  if (Array.isArray(value)) {
    let scalarTypes = value.reduce((acc, val) => {
      const type = getGraphQLType(val, potentialType)
      if (!acc.includes(type)) {
        acc.push(type)
      }
      return acc
    }, [])

    // Merge Float and Int together
    if (scalarTypes.includes('Float') && scalarTypes.includes('Int')) {
      scalarTypes = ['Float']
    }

    return `List<${scalarTypes.join(' | ')}>`
  }
  const jsType = typeof value
  if (jsType === 'number') {
    if (Math.trunc(value) === value) {
      return 'Int'
    } else {
      return 'Float'
    }
  }
  if (Object.prototype.toString.call(value) === '[object Date]') {
    return 'DateTime'
  }
  if (jsType === 'string') {
    if (
      /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(
        value,
      )
    ) {
      return 'UUID'
    }
    const date = new Date(value)
    if (
      potentialType &&
      typeof potentialType === 'object' &&
      (potentialType as DMMF.SchemaEnum).values &&
      (potentialType as DMMF.SchemaEnum).values.includes(value)
    ) {
      return potentialType.name
    }
    if (date.toString() === 'Invalid Date') {
      return 'String'
    }
    if (
      /\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)/.test(
        value,
      )
    ) {
      return 'DateTime'
    }
  }
  return JSTypeToGraphQLType[jsType]
}

export function graphQLToJSType(gql: string) {
  return GraphQLScalarToJSTypeTable[gql]
}

export function getSuggestion(
  str: string,
  possibilities: string[],
): string | null {
  const bestMatch = possibilities.reduce<{
    distance: number
    str: string | null
  }>(
    (acc, curr) => {
      const distance = leven(str, curr)
      if (distance < acc.distance) {
        return {
          distance,
          str: curr,
        }
      }

      return acc
    },
    {
      // heuristic to be not too strict, but allow some big mistakes (<= ~ 5)
      distance: Math.min(
        Math.floor(str.length) * 1.1,
        ...possibilities.map((p) => p.length * 3),
      ),
      str: null,
    },
  )

  return bestMatch.str
}

export function stringifyInputType(
  input: string | DMMF.InputType | DMMF.SchemaEnum,
  greenKeys: boolean = false,
): string {
  if (typeof input === 'string') {
    return input
  }
  if ((input as DMMF.SchemaEnum).values) {
    return `enum ${input.name} {\n${indent(
      (input as DMMF.SchemaEnum).values.join(', '),
      2,
    )}\n}`
  } else {
    const body = indent(
      (input as DMMF.InputType).fields // TS doesn't discriminate based on existence of fields properly
        .map((arg) => {
          const key = `${arg.name}`
          const str = `${greenKeys ? chalk.green(key) : key}${arg.isRequired ? '' : '?'
            }: ${chalk.white(
              arg.inputTypes
                .map((argType) => {
                  return wrapWithList(argIsInputType(argType.type)
                    ? argType.type.name
                    :
                    stringifyGraphQLType(argType.type),
                    argType.isList,
                  )
                }
                )
                .join(' | '),
            )}`
          if (!arg.isRequired) {
            return chalk.dim(str)
          }

          return str
        })
        .join('\n'),
      2,
    )
    return `${chalk.dim('type')} ${chalk.bold.dim(input.name)} ${chalk.dim(
      '{',
    )}\n${body}\n${chalk.dim('}')}`
  }
}

function argIsInputType(arg: DMMF.ArgType): arg is DMMF.InputType {
  if (typeof arg === 'string') {
    return false
  }

  return true
}

export function getInputTypeName(
  input: string | DMMF.InputType | DMMF.SchemaField | DMMF.SchemaEnum,
) {
  if (typeof input === 'string') {
    if (input === 'Null') {
      return 'null'
    }
    return input
  }

  return input.name
}

export function getOutputTypeName(
  input: string | DMMF.OutputType | DMMF.SchemaField | DMMF.SchemaEnum,
) {
  if (typeof input === 'string') {
    return input
  }

  return input.name
}

export function inputTypeToJson(
  input: string | DMMF.InputType | DMMF.SchemaEnum,
  isRequired: boolean,
  nameOnly: boolean = false,
): string | object {
  if (typeof input === 'string') {
    if (input === 'Null') {
      return 'null'
    }
    return input
  }

  if ((input as DMMF.SchemaEnum).values) {
    return (input as DMMF.SchemaEnum).values.join(' | ')
  }

  // TS "Trick" :/
  const inputType: DMMF.InputType = input as DMMF.InputType


  // If the parent type is required and all fields are non-scalars,
  // it's very useful to show to the user, which options they actually have
  const showDeepType =
    isRequired &&
    inputType.fields.every((arg) => arg.inputTypes[0].kind === 'object' || arg.inputTypes[1]?.kind === 'object')

  if (nameOnly) {
    return getInputTypeName(input)
  }

  return inputType.fields.reduce((acc, curr) => {
    let str = ''

    if (!showDeepType && !curr.isRequired) {
      str = curr.inputTypes.map(argType => getInputTypeName(argType.type)).join(' | ')
    } else {
      str = curr.inputTypes.map(argInputType => inputTypeToJson(argInputType.type, curr.isRequired, true)).join(' | ')
    }

    acc[curr.name + (curr.isRequired ? '' : '?')] = str
    return acc
  }, {})
}

export function destroyCircular(from, seen: any[] = []) {
  const to: any = Array.isArray(from) ? [] : {}

  seen.push(from)

  for (const key of Object.keys(from)) {
    const value = from[key]

    if (typeof value === 'function') {
      continue
    }

    if (!value || typeof value !== 'object') {
      to[key] = value
      continue
    }

    if (seen.indexOf(from[key]) === -1) {
      to[key] = destroyCircular(from[key], seen.slice(0))
      continue
    }

    to[key] = '[Circular]'
  }

  if (typeof from.name === 'string') {
    to.name = from.name
  }

  if (typeof from.message === 'string') {
    to.message = from.message
  }

  if (typeof from.stack === 'string') {
    to.stack = from.stack
  }

  return to
}

export function unionBy<T>(
  arr1: T[],
  arr2: T[],
  iteratee: (element: T) => string | number,
): T[] {
  const map = {}

  for (const element of arr1) {
    map[iteratee(element)] = element
  }

  for (const element of arr2) {
    const key = iteratee(element)
    if (!map[key]) {
      map[key] = element
    }
  }

  return Object.values(map)
}

export function uniqBy<T>(
  arr: T[],
  iteratee: (element: T) => string | number,
): T[] {
  const map = {}

  for (const element of arr) {
    map[iteratee(element)] = element
  }

  return Object.values(map)
}

export function capitalize(str: string): string {
  return str[0].toUpperCase() + str.slice(1)
}

/**
 * Converts the first character of a word to lower case
 * @param name
 */
export function lowerCase(name: string): string {
  return name.substring(0, 1).toLowerCase() + name.substring(1)
}
