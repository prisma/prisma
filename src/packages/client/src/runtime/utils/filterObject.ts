export function filterObject(obj, cb) {
  if (
    !obj ||
    typeof obj !== 'object' ||
    typeof obj.hasOwnProperty !== 'function'
  ) {
    return obj
  }
  const newObj = {}
  for (const key in obj) {
    const value = obj[key]
    if (obj.hasOwnProperty(key) && cb(key, value)) {
      newObj[key] = value
    }
  }
  return newObj
}
