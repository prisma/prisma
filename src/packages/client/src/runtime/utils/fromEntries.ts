export function fromEntries<T>(iterable: Array<[string, T]>): {
  [key: string]: T
} {
  return [...iterable].reduce<{ [key: string]: T }>((obj, [key, val]) => {
    obj[key] = val
    return obj
  }, {})
}
