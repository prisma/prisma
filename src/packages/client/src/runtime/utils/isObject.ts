/**
 * If we have a union type let's say of DateTimeFilter | Date, we will be called by tryInferArg.
 * And there are a few false friends out there, like `[object Date]` `[object BigInt]` or `Buffer`.
 * All of them have `typeof value === 'object'`, however, they're not objects in the Prisma sense
 */
const notReallyObjects = {
  '[object Date]': true,
  '[object BitInt]': true,
  '[object Uint8Array]': true, // for Buffers
  '[object Function]': true // for Decimal
}

export function isObject(value: any): boolean {
  return value
    && typeof value === 'object'
    && !notReallyObjects[Object.prototype.toString.call(value)]
}
