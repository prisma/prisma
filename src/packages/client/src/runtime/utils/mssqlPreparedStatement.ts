// Generate something like: SELECT * FROM User WHERE name = @P1 AND email = @P2 ...
export const mssqlPreparedStatement = (
  template:  ReadonlyArray<string> | TemplateStringsArray,
) => {
  return template.reduce((acc, str, idx) => `${acc}@P${idx}${str}`)
}
