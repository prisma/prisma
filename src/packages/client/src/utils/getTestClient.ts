import {
  extractPreviewFeatures,
  getConfig,
  getEnvPaths,
  getRelativeSchemaPath,
  mapPreviewFeatures,
  printConfigWarnings,
} from '@prisma/sdk'
import fs from 'fs'
import path from 'path'
import { parse } from 'stacktrace-parser'
import { promisify } from 'util'
import { extractSqliteSources } from '../generation/extractSqliteSources'
import { getDMMF } from '../generation/getDMMF'
import {
  getPrismaClient,
  GetPrismaClientOptions,
} from '../runtime/getPrismaClient'
import { generateInFolder } from './generateInFolder'
const readFile = promisify(fs.readFile)

/**
 * Returns an in-memory client for testing
 */
export async function getTestClient(
  schemaDir?: string,
  printWarnings?: boolean,
): Promise<any> {
  if (!schemaDir) {
    const callsite = parse(new Error('').stack!)
    schemaDir = path.dirname(callsite[1].file!)
  }
  const schemaPath = await getRelativeSchemaPath(schemaDir)
  const datamodel = await readFile(schemaPath!, 'utf-8')
  const config = await getConfig({ datamodel, ignoreEnvVarErrors: true })
  if (printWarnings) {
    printConfigWarnings(config.warnings)
  }

  const generator = config.generators.find(
    (g) => g.provider === 'prisma-client-js',
  )
  const enableExperimental = mapPreviewFeatures(extractPreviewFeatures(config))
  const document = await getDMMF({
    datamodel,
    enableExperimental,
  })
  const outputDir = schemaDir
  const relativeEnvPaths = getEnvPaths(schemaPath, { cwd: schemaDir })

  const options: GetPrismaClientOptions = {
    document,
    generator,
    dirname: schemaDir,
    relativePath: path.relative(outputDir, schemaDir),
    clientVersion: 'client-test-version',
    engineVersion: 'engine-test-version',
    relativeEnvPaths,
    datasourceNames: config.datasources.map((d) => d.name),
    sqliteDatasourceOverrides: extractSqliteSources(
      datamodel,
      schemaDir,
      outputDir,
    ),
    activeProvider: config.datasources[0].activeProvider,
  }

  return getPrismaClient(options)
}

/**
 * Actually generates a test client with its own query-engine into ./@prisma/client
 */
export async function generateTestClient(projectDir?: string): Promise<any> {
  if (!projectDir) {
    const callsite = parse(new Error('').stack!)
    projectDir = path.dirname(callsite[1].file!)
  }

  await generateInFolder({
    projectDir,
    useLocalRuntime: false,
    transpile: true,
    useBuiltRuntime: false,
  })
}
