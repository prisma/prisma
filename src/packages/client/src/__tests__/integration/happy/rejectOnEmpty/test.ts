import { getTestClient } from '../../../../utils/getTestClient'
const cases = {
  contructor: {
    customError: new Error('Contructor Custom Error'),
    customErrorPerModel: {
      User: new Error('Contructor Custom Error on User'),
    },
    true: true,
    false: false,
    undefined: undefined
  },
  findUnique: {
    customError: new Error('FindUnique Custom Error'),
    true: true,
    false: false,
    undefined: undefined
  }
}
for (const constructorKey of Object.keys(cases.contructor)) {
  const constructor = cases.contructor[constructorKey]
  for (const findUniqueKey of Object.keys(cases.findUnique)) {
    const  findUnique = cases.findUnique[ findUniqueKey]
    test(`rejectOnEmpty contructor=${constructorKey} findUnique=${findUniqueKey}`, async () => {
      // It should fail or not
      expect.assertions(1)
      const PrismaClient = await getTestClient()
      const prisma = new PrismaClient({
        rejectOnEmpty: constructor
      })

      // Test Rejection
      try {
        const r = await prisma.user.findUnique({
          where: {id: 'none'},
          rejectOnEmpty: findUnique
        })
        expect(r).toMatchSnapshot()
      } catch (error) {
        expect(error).toMatchSnapshot()
      }
      prisma.$disconnect()
    })
    
  }
}
