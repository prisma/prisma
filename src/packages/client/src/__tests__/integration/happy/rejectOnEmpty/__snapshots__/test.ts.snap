// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rejectOnEmpty contructor=customError findUnique=customError 1`] = `FindUnique Custom Error`;

exports[`rejectOnEmpty contructor=customError findUnique=false 1`] = `null`;

exports[`rejectOnEmpty contructor=customError findUnique=true 1`] = `No User found`;

exports[`rejectOnEmpty contructor=customError findUnique=undefined 1`] = `Contructor Custom Error`;

exports[`rejectOnEmpty contructor=customErrorPerModel findUnique=customError 1`] = `FindUnique Custom Error`;

exports[`rejectOnEmpty contructor=customErrorPerModel findUnique=false 1`] = `null`;

exports[`rejectOnEmpty contructor=customErrorPerModel findUnique=true 1`] = `No User found`;

exports[`rejectOnEmpty contructor=customErrorPerModel findUnique=undefined 1`] = `Contructor Custom Error on User`;

exports[`rejectOnEmpty contructor=false findUnique=customError 1`] = `FindUnique Custom Error`;

exports[`rejectOnEmpty contructor=false findUnique=false 1`] = `null`;

exports[`rejectOnEmpty contructor=false findUnique=true 1`] = `No User found`;

exports[`rejectOnEmpty contructor=false findUnique=undefined 1`] = `null`;

exports[`rejectOnEmpty contructor=true findUnique=customError 1`] = `FindUnique Custom Error`;

exports[`rejectOnEmpty contructor=true findUnique=false 1`] = `null`;

exports[`rejectOnEmpty contructor=true findUnique=true 1`] = `No User found`;

exports[`rejectOnEmpty contructor=true findUnique=undefined 1`] = `No User found`;

exports[`rejectOnEmpty contructor=undefined findUnique=customError 1`] = `FindUnique Custom Error`;

exports[`rejectOnEmpty contructor=undefined findUnique=false 1`] = `null`;

exports[`rejectOnEmpty contructor=undefined findUnique=true 1`] = `No User found`;

exports[`rejectOnEmpty contructor=undefined findUnique=undefined 1`] = `null`;
